### PR TITLE
First rough pass at enemy as hookable

### PR DESCRIPTION
**Pull request by Razoric480**
Wanted to get a very basic example going for review, with as little changes to player as possible.

It doesn't check all the boxes of the tasks (enemy doesn't damage player, nor does it get knocked back) but what is currently in there was done with no change to Player states. Thought it'd be worth a look to make sure it's not doing any no-no or falling into a pitfall I didn't see coming.

### CODE REVIEW

**Comment by NathanLovato** [comment]
Nice start. Regarding the jump it's not what I had in mind:

- Hooking on the enemy is what damages the enemy
- Jumping/using the enemy as a bumper is an automated animation, the player doesn't have to jump on the enemy to kill it, to damage it, or do anything.

In other words, when the player hooks onto the enemy:

1. It moves to the enemy like a normal hooking point, and the hook itself damages the enemy
2. The player automatically uses the enemy as a stepping stone to jump or hop vertically

Number 2. is so that the player has a bit of time in the air to aim and hook onto another target.

If it's not clear, I can tweak the prototype to show the idea

**Comment by Razoric480** [comment]
Changed it up with your revision. Hope that's closer to what you had in mind.

**Comment by NathanLovato** [comment]
One step closer, but not there yet. I'll do the next iteration to show what I mean, as it's a matter of game design and game feel. The way the character hooks jumps and jumps now gives no sense of impact or combat in general, it would also make it hard to aim for the next enemy

**Comment by NathanLovato** [comment]
There you go, I added a small commit to show you the idea: the player stops on the enemy's back for an instant, then jump. The idea would be to have a nicer animation/interaction between the two, but at least that's a start.

Code-wise, I don't know how we could best handle having the player propel himself from the enemy's body, i.e. knocking back the enemy when the player does that jump. One option would be to pass the enemy to the player's HopOnEnemy state, but I'd rather not control the enemy from the player's code and vice-versa.
One thing that could work would be a one-time signal connection on the enemy, to make it react to the player having hopped.

**Comment by NathanLovato** [comment]
The changes are here: https://github.com/GDquest/godot-platformer-2d/pull/163/commits/113f83e35824e43f97b12f39290c464fac34790f

**Comment by Razoric480** [comment]
I see! Okay, I get it now. I'll work with it from that angle.

**Comment by Razoric480** [comment]
Moved whether the player state goes into HopOnEnemy out of the enemy, and instead having the Hook state of the player object transition to HopOnEnemy is the hook target was a KinematicBody (otherwise transition to Move/Air as normal.)

It seems to play pretty alright, and keeps the code from infecting one another, but I'll let you judge.

File: game/src/AI/States/Destroyed.gd [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+func enter(msg: Dictionary = {}) -> void:
+	owner.hook_target.set_color(Color(0, 0, 0, 0))
+	owner.body.set_color_fill(hurt_color)
+	
+	yield(do_color(), "completed")
+	
+	owner.queue_free()
+
+
+func do_color() -> void:
```
Comment (NathanLovato) [inline game/src/AI/States/Destroyed.gd]: Use an animation player for things like these, you're making your life harder than it needs to be, and in general the anim player will likely have better perfs than any GDScript code

File: game/src/AI/States/Hooked.gd [hunk 2]
```diff
@@ -0,0 +1,19 @@
+extends State
+"""
+This state is for enemy's response to being grappled with the hook.
+It simply waits until the player body collides with the hitbox, colors the body,
+ then transitions to `state_when_struck`
+"""
+
+
+export var state_when_struck: String
```
Comment (NathanLovato) [inline game/src/AI/States/Hooked.gd]: Do we really want to type the path to the state in a variable? Does it ever need to be "hard-coded"? The case I can see it change is e.g. going to Stagger/Knockback if the enemy doesn't die with a given hit vs Die if it dies. But that's not something we should change with a property, I'd add that with the `msg` dictionary when the need arises

File: game/src/AI/States/HorizontalPatrol.gd [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+
+
+onready var _move_speed: float = owner.move_speed
+onready var _direction: int = owner.direction
+
+
+func enter(msg: Dictionary = {}) -> void:
+	owner.hook_target.connect("hooked_onto_from", self, "_on_Hook_hooked_onto_from", [], CONNECT_ONESHOT)
+	owner.hitbox.connect("body_entered", self, "_on_Player_body_entered")
+	owner.hook_target.set_is_active(true)
```
Comment (NathanLovato) [inline game/src/AI/States/HorizontalPatrol.gd]: We use properties over setters. `owner.hook_target.is_active = true`

**Comment by NathanLovato** [comment]
I left a few code style/guideline-related comments, but in general it's looking good. To me, the only thing left for this pass is making the player temporarily invincible after killing an enemy: if there's an enemy above the player, the jump after hooking onto an enemy will kill you.

But that may be a task for another branch because I think this should interact with the health system. Health should be external to the state machine and handle everything related to health and invincibility.

**Comment by NathanLovato** [comment]
So if you think it is good to merge now, we can merge this first pass and add that later. 🙂

**Pull request merged**